File: src/core/filter.ts
import { stripAnsi } from './utils';

export enum FilterLevel {
  None = 'none',
  Minimal = 'minimal',
  Aggressive = 'aggressive',
}

const LANG_BY_EXT: Record<string, string> = {
  ts: 'ts', tsx: 'tsx',
  js: 'js', jsx: 'jsx', mjs: 'js', cjs: 'js',
  py: 'py',
  go: 'go',
  rs: 'rs',
  java: 'java',
  cs: 'cs',
  cpp: 'cpp', cc: 'cpp', cxx: 'cpp', hpp: 'cpp', hxx: 'cpp',
  c: 'c', h: 'c',
  rb: 'rb',
};

export function langFromPath(p: string): string {
  const ext = p.split('.').pop()?.toLowerCase() || '';
  return LANG_BY_EXT[ext] || '';
}

function stripBlockComments(src: string): string {
  return src.replace(/\/\*[\s\S]*?\*\//g, '');
}

function stripLineComments(src: string, prefix: string): string {
  return src
    .split('\n')
    .map((line) => {
      const idx = findCommentIdx(line, prefix);
      return idx >= 0 ? line.slice(0, idx).replace(/\s+$/, '') : line;
    })
    .join('\n');
}

function findCommentIdx(line: string, prefix: string): number {
  let inSingle = false;
  let inDouble = false;
  let inBacktick = false;
  for (let i = 0; i < line.length; i++) {
    const c = line[i];
    const prev = i > 0 ? line[i - 1] : '';
    if (c === "'" && prev !== '\\' && !inDouble && !inBacktick) inSingle = !inSingle;
    else if (c === '"' && prev !== '\\' && !inSingle && !inBacktick) inDouble = !inDouble;
    else if (c === '`' && prev !== '\\' && !inSingle && !inDouble) inBacktick = !inBacktick;
    else if (!inSingle && !inDouble && !inBacktick) {
      if (line.startsWith(prefix, i)) return i;
    }
  }
  return -1;
}

function collapseBlanks(src: string): string {
  return src.replace(/\n{3,}/g, '\n\n');
}

const C_LIKE = new Set(['ts', 'tsx', 'js', 'jsx', 'go', 'rs', 'java', 'cs', 'cpp', 'c']);
const HASH_LANG = new Set(['py', 'rb']);

export function filterCode(source: string, lang: string, level: FilterLevel): string {
  if (level === FilterLevel.None) return source;

  let out = source;
  if (C_LIKE.has(lang)) {
    out = stripBlockComments(out);
    out = stripLineComments(out, '//');
  } else if (HASH_LANG.has(lang)) {
    out = stripLineComments(out, '#');
  } else {
    out = stripBlockComments(out);
    out = stripLineComments(out, '//');
    out = stripLineComments(out, '#');
  }

  out = collapseBlanks(out);

  if (level === FilterLevel.Minimal) return out;

  return aggressiveFilter(out, lang);
}

function aggressiveFilter(src: string, lang: string): string {
  if (C_LIKE.has(lang)) return aggressiveCLike(src);
  if (lang === 'py') return aggressivePython(src);
  if (lang === 'rb') return aggressiveRuby(src);
  return src;
}

function aggressiveCLike(src: string): string {
  const out: string[] = [];
  const lines = src.split('\n');
  let i = 0;
  while (i < lines.length) {
    const line = lines[i];
    const trimmed = line.trim();
    const keepers = /^(import\b|export\b|from\b|const\b|let\b|var\b|type\b|interface\b|enum\b|namespace\b|declare\b|use\b|using\b|package\b)/;
    const sigStart = /^(export\s+)?(default\s+)?(async\s+)?(function|class|interface|type|enum)\b/;
    const methodSig = /^[a-zA-Z_$][\w$]*\s*\([^)]*\)\s*(:\s*[^{]+)?\s*\{?\s*$/;
    const arrowFn = /=>\s*\{?\s*$/;

    if (sigStart.test(trimmed) && trimmed.includes('{')) {
      const end = findMatchingBrace(lines, i, line.indexOf('{'));
      const sig = line.replace(/\{[^]*$/, '{ ... }');
      out.push(sig);
      i = end + 1;
      continue;
    }

    if (sigStart.test(trimmed) && !trimmed.includes('{')) {
      out.push(line);
      i++;
      while (i < lines.length && !lines[i].includes('{') && !lines[i].endsWith(';')) {
        out.push(lines[i]);
        i++;
      }
      if (i < lines.length && lines[i].includes('{')) {
        const end = findMatchingBrace(lines, i, lines[i].indexOf('{'));
        out.push(lines[i].replace(/\{[^]*$/, '{ ... }'));
        i = end + 1;
        continue;
      }
      continue;
    }

    if (methodSig.test(trimmed) && trimmed.includes('{')) {
      const end = findMatchingBrace(lines, i, line.indexOf('{'));
      out.push(line.replace(/\{[^]*$/, '{ ... }'));
      i = end + 1;
      continue;
    }

    if (arrowFn.test(trimmed) && trimmed.includes('{')) {
      const end = findMatchingBrace(lines, i, line.lastIndexOf('{'));
      out.push(line.replace(/\{[^]*$/, '{ ... }'));
      i = end + 1;
      continue;
    }

    if (keepers.test(trimmed)) {
      out.push(line);
      i++;
      continue;
    }

    if (trimmed === '' || /^[}\])]+;?\s*$/.test(trimmed)) {
      out.push(line);
      i++;
      continue;
    }

    i++;
  }
  return collapseBlanks(out.join('\n'));
}

function findMatchingBrace(lines: string[], startLine: number, startCol: number): number {
  let depth = 0;
  let started = false;
  for (let li = startLine; li < lines.length; li++) {
    const line = lines[li];
    const from = li === startLine ? startCol : 0;
    for (let ci = from; ci < line.length; ci++) {
      const c = line[ci];
      if (c === '{') {
        depth++;
        started = true;
      } else if (c === '}') {
        depth--;
        if (started && depth === 0) return li;
      }
    }
  }
  return lines.length - 1;
}

function aggressivePython(src: string): string {
  const lines = src.split('\n');
  const out: string[] = [];
  const sig = /^(\s*)(async\s+)?(def|class)\s+/;
  let i = 0;
  while (i < lines.length) {
    const line = lines[i];
    const m = sig.exec(line);
    if (m) {
      const baseIndent = m[1].length;
      out.push(`${line.replace(/:\s*$/, ':')} ...`);
      i++;
      while (i < lines.length) {
        const next = lines[i];
        const trimmed = next.trim();
        if (trimmed === '') { i++; continue; }
        const indent = next.search(/\S/);
        if (indent <= baseIndent) break;
        i++;
      }
      continue;
    }
    if (/^(import |from )/.test(line) || /^[A-Z_]+\s*=/.test(line.trim())) {
      out.push(line);
    } else if (line.trim() === '') {
      out.push(line);
    }
    i++;
  }
  return collapseBlanks(out.join('\n'));
}

function aggressiveRuby(src: string): string {
  const lines = src.split('\n');
  const out: string[] = [];
  let i = 0;
  while (i < lines.length) {
    const line = lines[i];
    const trimmed = line.trim();
    if (/^(def |class |module )/.test(trimmed)) {
      out.push(line);
      i++;
      while (i < lines.length && !/^\s*end\b/.test(lines[i])) {
        i++;
      }
      if (i < lines.length) {
        out.push(lines[i]);
        i++;
      }
      continue;
    }
    if (/^(require|include|extend|attr_)/.test(trimmed) || trimmed === '') {
      out.push(line);
    }
    i++;
  }
  return collapseBlanks(out.join('\n'));
}

export function smartSummary(source: string, lang: string): string {
  const clean = stripAnsi(source);
  const lines = clean.split('\n');
  const lineCount = lines.length;

  if (lang === 'tsx' || lang === 'jsx') {
    const props = countMatches(clean, /interface\s+\w*Props|type\s+\w*Props\s*=|\bprops\.\w+/g);
    const hooks = countMatches(clean, /\buse[A-Z]\w+\s*\(/g);
    const isComponent = /export\s+(default\s+)?(function|const)\s+[A-Z]/.test(clean) ||
      /\breturn\s*\(\s*</.test(clean);
    const renders = isComponent ? 'a component' : 'JSX content';
    return `React component: ${props} props, ${hooks} hooks, renders ${renders}\n${lineCount} lines total`;
  }

  if (lang === 'ts' || lang === 'js') {
    const classes = countMatches(clean, /\bclass\s+\w+/g);
    const interfaces = countMatches(clean, /\binterface\s+\w+/g);
    const exports = countMatches(clean, /\bexport\s+(default\s+|const\s+|function\s+|class\s+|interface\s+|type\s+|enum\s+|let\s+|var\s+)/g);
    const functions = countMatches(clean, /\bfunction\s+\w+|\bconst\s+\w+\s*=\s*(async\s+)?\(/g);
    const asyncFns = countMatches(clean, /\basync\s+(function|\()/g);
    const imports = countMatches(clean, /^import\s+/gm);
    if (classes > 0) {
      return `TypeScript class: ${classes} classes, ${interfaces} interfaces, ${imports} imports\n${functions} functions (${asyncFns} async), ${lineCount} lines`;
    }
    return `Node module: ${exports} exports, ${asyncFns} async functions, ${imports} imports\n${functions} functions, ${lineCount} lines`;
  }

  if (lang === 'py') {
    const classes = countMatches(clean, /^\s*class\s+\w+/gm);
    const defs = countMatches(clean, /^\s*(async\s+)?def\s+\w+/gm);
    const imports = countMatches(clean, /^(import\s+|from\s+\w)/gm);
    return `Python module: ${classes} classes, ${defs} functions, ${imports} imports\n${lineCount} lines total`;
  }

  if (lang === 'go') {
    const funcs = countMatches(clean, /^func\s+/gm);
    const types = countMatches(clean, /^type\s+\w+/gm);
    const imports = countMatches(clean, /^import\s+/gm);
    return `Go file: ${funcs} functions, ${types} types, ${imports} import blocks\n${lineCount} lines total`;
  }

  return `${lang || 'text'} file: ${lineCount} lines, ${clean.length} bytes`;
}

function countMatches(s: string, re: RegExp): number {
  const m = s.match(re);
  return m ? m.length : 0;
}

export function deduplicateLines(raw: string): string {
  const lines = stripAnsi(raw).split('\n');
  const order: string[] = [];
  const counts = new Map<string, { display: string; count: number }>();

  for (const line of lines) {
    const normalized = line
      .replace(/\d{4}-\d{2}-\d{2}[T ]\d{2}:\d{2}:\d{2}(\.\d+)?(Z|[+-]\d{2}:?\d{2})?/g, '<TS>')
      .replace(/\b[0-9a-f]{8,}\b/gi, '<ID>')
      .trim();
    if (normalized === '') continue;
    const existing = counts.get(normalized);
    if (existing) {
      existing.count++;
    } else {
      counts.set(normalized, { display: line, count: 1 });
      order.push(normalized);
    }
  }

  const sorted = order
    .map((k) => counts.get(k)!)
    .sort((a, b) => b.count - a.count);

  return sorted
    .map((entry) => (entry.count >= 2 ? `${entry.display} (×${entry.count})` : entry.display))
    .join('\n');
}

export function extractStructure(value: unknown, depth = 0): unknown {
  if (depth > 6) return '<...>';
  if (value === null) return 'null';
  if (Array.isArray(value)) {
    if (value.length === 0) return [];
    return [extractStructure(value[0], depth + 1)];
  }
  const t = typeof value;
  if (t === 'object') {
    const obj: Record<string, unknown> = {};
    for (const [k, v] of Object.entries(value as Record<string, unknown>)) {
      obj[k] = extractStructure(v, depth + 1);
    }
    return obj;
  }
  return t;
}

File: internal/filter/filter.go
// Package filter holds tok's compression primitives: language-aware source stripping,
// output line dedup, and JSON structure extraction. Pure functions, no I/O.
package filter

import (
	"fmt"
	"regexp"
	"sort"
	"strings"

	"github.com/Kevalgor12/tok-proxy/internal/config"
	"github.com/Kevalgor12/tok-proxy/internal/util"
)

var langByExt = map[string]string{
	"ts": "ts", "tsx": "tsx",
	"js": "js", "jsx": "jsx", "mjs": "js", "cjs": "js",
	"py":   "py",
	"go":   "go",
	"rs":   "rs",
	"java": "java",
	"cs":   "cs",
	"cpp":  "cpp", "cc": "cpp", "cxx": "cpp", "hpp": "cpp", "hxx": "cpp",
	"c": "c", "h": "c",
	"rb": "rb",
}

func LangFromPath(p string) string {
	dot := strings.LastIndex(p, ".")
	if dot < 0 {
		return ""
	}
	return langByExt[strings.ToLower(p[dot+1:])]
}

var (
	cLike    = map[string]bool{"ts": true, "tsx": true, "js": true, "jsx": true, "go": true, "rs": true, "java": true, "cs": true, "cpp": true, "c": true}
	hashLang = map[string]bool{"py": true, "rb": true}

	blockCommentRe = regexp.MustCompile(`(?s)/\*.*?\*/`)
	blankRunRe     = regexp.MustCompile(`\n{3,}`)
)

// FilterCode strips comments (and, at the aggressive level, function bodies) from source.
func FilterCode(source, lang string, level config.FilterLevel) string {
	if level == config.FilterNone {
		return source
	}

	out := source
	switch {
	case cLike[lang]:
		out = blockCommentRe.ReplaceAllString(out, "")
		out = stripLineComments(out, "//")
	case hashLang[lang]:
		out = stripLineComments(out, "#")
	default:
		out = blockCommentRe.ReplaceAllString(out, "")
		out = stripLineComments(out, "//")
		out = stripLineComments(out, "#")
	}
	out = collapseBlanks(out)

	if level == config.FilterMinimal {
		return out
	}
	return aggressiveFilter(out, lang)
}

func stripLineComments(src, prefix string) string {
	lines := strings.Split(src, "\n")
	for i, line := range lines {
		if idx := findCommentIdx(line, prefix); idx >= 0 {
			lines[i] = strings.TrimRight(line[:idx], " \t\r\f\v")
		}
	}
	return strings.Join(lines, "\n")
}

// findCommentIdx returns the column where `prefix` begins outside any string literal,
// or -1. Scans bytes: string delimiters and comment markers are all ASCII, so multi-byte
// runes elsewhere on the line can't be mistaken for them.
func findCommentIdx(line, prefix string) int {
	var inSingle, inDouble, inBacktick bool
	for i := 0; i < len(line); i++ {
		c := line[i]
		var prev byte
		if i > 0 {
			prev = line[i-1]
		}
		switch {
		case c == '\'' && prev != '\\' && !inDouble && !inBacktick:
			inSingle = !inSingle
		case c == '"' && prev != '\\' && !inSingle && !inBacktick:
			inDouble = !inDouble
		case c == '`' && prev != '\\' && !inSingle && !inDouble:
			inBacktick = !inBacktick
		case !inSingle && !inDouble && !inBacktick:
			if strings.HasPrefix(line[i:], prefix) {
				return i
			}
		}
	}
	return -1
}

func collapseBlanks(src string) string {
	return blankRunRe.ReplaceAllString(src, "\n\n")
}

func aggressiveFilter(src, lang string) string {
	switch {
	case cLike[lang]:
		return aggressiveCLike(src)
	case lang == "py":
		return aggressivePython(src)
	case lang == "rb":
		return aggressiveRuby(src)
	default:
		return src
	}
}

var (
	keepersRe   = regexp.MustCompile(`^(import\b|export\b|from\b|const\b|let\b|var\b|type\b|interface\b|enum\b|namespace\b|declare\b|use\b|using\b|package\b)`)
	sigStartRe  = regexp.MustCompile(`^(export\s+)?(default\s+)?(async\s+)?(function|class|interface|type|enum)\b`)
	methodSigRe = regexp.MustCompile(`^[a-zA-Z_$][\w$]*\s*\([^)]*\)\s*(:\s*[^{]+)?\s*\{?\s*$`)
	arrowFnRe   = regexp.MustCompile(`=>\s*\{?\s*$`)
	closerRe    = regexp.MustCompile(`^[}\])]+;?\s*$`)
)

func aggressiveCLike(src string) string {
	lines := strings.Split(src, "\n")
	var out []string
	i := 0
	for i < len(lines) {
		line := lines[i]
		trimmed := strings.TrimSpace(line)

		if sigStartRe.MatchString(trimmed) && strings.Contains(trimmed, "{") {
			end := findMatchingBrace(lines, i, strings.Index(line, "{"))
			out = append(out, collapseToBraceStub(line))
			i = end + 1
			continue
		}

		if sigStartRe.MatchString(trimmed) && !strings.Contains(trimmed, "{") {
			out = append(out, line)
			i++
			for i < len(lines) && !strings.Contains(lines[i], "{") && !strings.HasSuffix(lines[i], ";") {
				out = append(out, lines[i])
				i++
			}
			if i < len(lines) && strings.Contains(lines[i], "{") {
				end := findMatchingBrace(lines, i, strings.Index(lines[i], "{"))
				out = append(out, collapseToBraceStub(lines[i]))
				i = end + 1
			}
			continue
		}

		if methodSigRe.MatchString(trimmed) && strings.Contains(trimmed, "{") {
			end := findMatchingBrace(lines, i, strings.Index(line, "{"))
			out = append(out, collapseToBraceStub(line))
			i = end + 1
			continue
		}

		if arrowFnRe.MatchString(trimmed) && strings.Contains(trimmed, "{") {
			end := findMatchingBrace(lines, i, strings.LastIndex(line, "{"))
			out = append(out, collapseToBraceStub(line))
			i = end + 1
			continue
		}

		if keepersRe.MatchString(trimmed) {
			out = append(out, line)
			i++
			continue
		}

		if trimmed == "" || closerRe.MatchString(trimmed) {
			out = append(out, line)
			i++
			continue
		}

		i++
	}
	return collapseBlanks(strings.Join(out, "\n"))
}

// collapseToBraceStub replaces everything from the first `{` to end of line with `{ ... }`.
func collapseToBraceStub(line string) string {
	if i := strings.Index(line, "{"); i >= 0 {
		return line[:i] + "{ ... }"
	}
	return line
}

func findMatchingBrace(lines []string, startLine, startCol int) int {
	depth := 0
	started := false
	for li := startLine; li < len(lines); li++ {
		line := lines[li]
		from := 0
		if li == startLine {
			from = startCol
		}
		for ci := from; ci < len(line); ci++ {
			switch line[ci] {
			case '{':
				depth++
				started = true
			case '}':
				depth--
				if started && depth == 0 {
					return li
				}
			}
		}
	}
	return len(lines) - 1
}

var (
	pySigRe    = regexp.MustCompile(`^(\s*)(async\s+)?(def|class)\s+`)
	pyColonRe  = regexp.MustCompile(`:\s*$`)
	pyImportRe = regexp.MustCompile(`^(import |from )`)
	pyConstRe  = regexp.MustCompile(`^[A-Z_]+\s*=`)
	nonSpaceRe = regexp.MustCompile(`\S`)
)

func aggressivePython(src string) string {
	lines := strings.Split(src, "\n")
	var out []string
	i := 0
	for i < len(lines) {
		line := lines[i]
		if m := pySigRe.FindStringSubmatch(line); m != nil {
			baseIndent := len(m[1])
			out = append(out, pyColonRe.ReplaceAllString(line, ":")+" ...")
			i++
			for i < len(lines) {
				next := lines[i]
				if strings.TrimSpace(next) == "" {
					i++
					continue
				}
				loc := nonSpaceRe.FindStringIndex(next)
				indent := len(next)
				if loc != nil {
					indent = loc[0]
				}
				if indent <= baseIndent {
					break
				}
				i++
			}
			continue
		}
		if pyImportRe.MatchString(line) || pyConstRe.MatchString(strings.TrimSpace(line)) {
			out = append(out, line)
		} else if strings.TrimSpace(line) == "" {
			out = append(out, line)
		}
		i++
	}
	return collapseBlanks(strings.Join(out, "\n"))
}

var (
	rbSigRe  = regexp.MustCompile(`^(def |class |module )`)
	rbEndRe  = regexp.MustCompile(`^\s*end\b`)
	rbKeepRe = regexp.MustCompile(`^(require|include|extend|attr_)`)
)

func aggressiveRuby(src string) string {
	lines := strings.Split(src, "\n")
	var out []string
	i := 0
	for i < len(lines) {
		line := lines[i]
		trimmed := strings.TrimSpace(line)
		if rbSigRe.MatchString(trimmed) {
			out = append(out, line)
			i++
			for i < len(lines) && !rbEndRe.MatchString(lines[i]) {
				i++
			}
			if i < len(lines) {
				out = append(out, lines[i])
				i++
			}
			continue
		}
		if rbKeepRe.MatchString(trimmed) || trimmed == "" {
			out = append(out, line)
		}
		i++
	}
	return collapseBlanks(strings.Join(out, "\n"))
}

// ---- summaries + dedup + structure -----------------------------------------

var (
	reTsxProps   = regexp.MustCompile(`interface\s+\w*Props|type\s+\w*Props\s*=|\bprops\.\w+`)
	reHooks      = regexp.MustCompile(`\buse[A-Z]\w+\s*\(`)
	reComponent  = regexp.MustCompile(`export\s+(default\s+)?(function|const)\s+[A-Z]`)
	reReturnJSX  = regexp.MustCompile(`\breturn\s*\(\s*<`)
	reClass      = regexp.MustCompile(`\bclass\s+\w+`)
	reInterface  = regexp.MustCompile(`\binterface\s+\w+`)
	reExport     = regexp.MustCompile(`\bexport\s+(default\s+|const\s+|function\s+|class\s+|interface\s+|type\s+|enum\s+|let\s+|var\s+)`)
	reFunction   = regexp.MustCompile(`\bfunction\s+\w+|\bconst\s+\w+\s*=\s*(async\s+)?\(`)
	reAsync      = regexp.MustCompile(`\basync\s+(function|\()`)
	reImportLine = regexp.MustCompile(`(?m)^import\s+`)
	rePyClass    = regexp.MustCompile(`(?m)^\s*class\s+\w+`)
	rePyDef      = regexp.MustCompile(`(?m)^\s*(async\s+)?def\s+\w+`)
	rePyImport   = regexp.MustCompile(`(?m)^(import\s+|from\s+\w)`)
	reGoFunc     = regexp.MustCompile(`(?m)^func\s+`)
	reGoType     = regexp.MustCompile(`(?m)^type\s+\w+`)
	reGoImport   = regexp.MustCompile(`(?m)^import\s+`)
)

func SmartSummary(source, lang string) string {
	clean := util.StripAnsi(source)
	lineCount := len(strings.Split(clean, "\n"))

	switch {
	case lang == "tsx" || lang == "jsx":
		props := count(clean, reTsxProps)
		hooks := count(clean, reHooks)
		isComponent := reComponent.MatchString(clean) || reReturnJSX.MatchString(clean)
		renders := "JSX content"
		if isComponent {
			renders = "a component"
		}
		return fmt.Sprintf("React component: %d props, %d hooks, renders %s\n%d lines total", props, hooks, renders, lineCount)

	case lang == "ts" || lang == "js":
		classes := count(clean, reClass)
		interfaces := count(clean, reInterface)
		exports := count(clean, reExport)
		functions := count(clean, reFunction)
		asyncFns := count(clean, reAsync)
		imports := count(clean, reImportLine)
		if classes > 0 {
			return fmt.Sprintf("TypeScript class: %d classes, %d interfaces, %d imports\n%d functions (%d async), %d lines", classes, interfaces, imports, functions, asyncFns, lineCount)
		}
		return fmt.Sprintf("Node module: %d exports, %d async functions, %d imports\n%d functions, %d lines", exports, asyncFns, imports, functions, lineCount)

	case lang == "py":
		return fmt.Sprintf("Python module: %d classes, %d functions, %d imports\n%d lines total", count(clean, rePyClass), count(clean, rePyDef), count(clean, rePyImport), lineCount)

	case lang == "go":
		return fmt.Sprintf("Go file: %d functions, %d types, %d import blocks\n%d lines total", count(clean, reGoFunc), count(clean, reGoType), count(clean, reGoImport), lineCount)

	default:
		name := lang
		if name == "" {
			name = "text"
		}
		return fmt.Sprintf("%s file: %d lines, %d bytes", name, lineCount, len(clean))
	}
}

func count(s string, re *regexp.Regexp) int { return len(re.FindAllString(s, -1)) }

var (
	tsRe = regexp.MustCompile(`\d{4}-\d{2}-\d{2}[T ]\d{2}:\d{2}:\d{2}(\.\d+)?(Z|[+-]\d{2}:?\d{2})?`)
	idRe = regexp.MustCompile(`(?i)\b[0-9a-f]{8,}\b`)
)

// DeduplicateLines collapses repeated log lines (ignoring timestamps and hex IDs) into a
// single line with an occurrence count, ordered most-frequent first.
func DeduplicateLines(raw string) string {
	lines := strings.Split(util.StripAnsi(raw), "\n")

	type entry struct {
		display string
		count   int
	}
	var order []string
	counts := map[string]*entry{}

	for _, line := range lines {
		normalized := idRe.ReplaceAllString(tsRe.ReplaceAllString(line, "<TS>"), "<ID>")
		normalized = strings.TrimSpace(normalized)
		if normalized == "" {
			continue
		}
		if e, ok := counts[normalized]; ok {
			e.count++
		} else {
			counts[normalized] = &entry{display: line, count: 1}
			order = append(order, normalized)
		}
	}

	entries := make([]*entry, len(order))
	for i, k := range order {
		entries[i] = counts[k]
	}
	sort.SliceStable(entries, func(i, j int) bool { return entries[i].count > entries[j].count })

	out := make([]string, len(entries))
	for i, e := range entries {
		if e.count >= 2 {
			out[i] = fmt.Sprintf("%s (×%d)", e.display, e.count)
		} else {
			out[i] = e.display
		}
	}
	return strings.Join(out, "\n")
}

// ExtractStructure reduces a decoded JSON value to its shape: arrays keep one sample
// element, objects keep their keys, and scalars become their type name.
func ExtractStructure(value any) any { return extractStructure(value, 0) }

func extractStructure(value any, depth int) any {
	if depth > 6 {
		return "<...>"
	}
	switch v := value.(type) {
	case nil:
		return "null"
	case []any:
		if len(v) == 0 {
			return []any{}
		}
		return []any{extractStructure(v[0], depth+1)}
	case map[string]any:
		out := map[string]any{}
		for k, val := range v {
			out[k] = extractStructure(val, depth+1)
		}
		return out
	case string:
		return "string"
	case float64:
		return "number"
	case bool:
		return "boolean"
	default:
		return "object"
	}
}
